### PR TITLE
Spark 3.1.0 APIs - DataFrame

### PR DESCRIPTION
**Pull request by suhsteve**
PR adds support for the following APIs added in Spark 3.1.0:

**DataFrame**
```scala
def unionByName(other: Dataset[T], allowMissingColumns: Boolean): Dataset[T]
def sameSemantics(other: Dataset[T]): Boolean
def semanticHash(): Int
```

Added into scala in Spark 2.0.0, but added into pySpark in Spark 3.1.0
```scala
def inputFiles: Array[String]
```

### CODE REVIEW

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs [hunk 1]
```diff
@@ -696,10 +696,10 @@ public void TestSignaturesV2_4_X()
         }
 
         /// <summary>
-        /// Test signatures for APIs introduced in Spark 3.*
+        /// Test signatures for APIs introduced in Spark 3.0.*
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs:699]:
```suggestion
        /// Test signatures for APIs introduced in Spark 3.0.*.

```

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs [hunk 2]
```diff
@@ -729,5 +729,18 @@ public void TestSignaturesV3_X_X()
             _df.Explain("cost");
             _df.Explain("formatted");
         }
+
+        /// <summary>
+        /// Test signatures for APIs introduced in Spark 3.1.*
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs:734]:
```suggestion
        /// Test signatures for APIs introduced in Spark 3.1.*.

```

**Comment by Niharikadutta** [comment]
LGTM @suhsteve (few minor nits). Thanks!

**Comment by Niharikadutta** [comment]
@suhsteve One quick question, should we include the `inputFiles` API as introduced in [Python from 3.1.0](https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/python/pyspark/sql/dataframe.py#L2626)?

**Comment by suhsteve** [comment]
> @suhsteve One quick question, should we include the `inputFiles` API as introduced in [Python from 3.1.0](https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/python/pyspark/sql/dataframe.py#L2626)?

Hmm, that API was introduced on the scala side since 2.0.0, but just got into pyspark in 3.1.0.  Wouldn't hurt to expose it as well.

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs [hunk 3]
```diff
@@ -729,5 +732,19 @@ public void TestSignaturesV3_X_X()
             _df.Explain("cost");
             _df.Explain("formatted");
         }
+
+        /// <summary>
+        /// Test signatures for APIs introduced in Spark 3.1.*.
+
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs:738]:
```suggestion

```

**Review by imback82 — DISMISSED** [review]
LGTM, thanks @suhsteve!

**Pull request merged**